include/boost/iostreams/filter/zlib.hpp

```diff
@@ -347,7 +347,7 @@ bool zlib_compressor_impl<Alloc>::filter
     int result = xdeflate(flush ? zlib::finish : zlib::no_flush);
     after(src_begin, dest_begin, true);
     zlib_error::check BOOST_PREVENT_MACRO_SUBSTITUTION(result);
-    return result != zlib::stream_end; 
+    return result == zlib::okay; 
```

Comment:
It's not always a user-buffer error, conversely it's an internal buffer error usually. You can get this error with a very big user buffer. It's meaningless to throw an exception for an internal recoverable error. I use this library intensively for io operations with high throughput, and this fix really works for Z_BUF_ERROR conditions.